tools/checkstyle/checkstyle.bzl

```diff
@@ -0,0 +1,199 @@
+                if f.extension == 'java':
+                    files.append(f)
+    return [JavaSourceFiles(files = files)]
+
+
+collect_sources = aspect(
+    implementation = collect_sources_impl,
+)
+
+
+# `ctx` is rule context: https://docs.bazel.build/versions/master/skylark/lib/ctx.html
+def _checkstyle_test_impl(ctx):
+    # verify target name matches naming conventions
+    if "{}-checkstyle".format(ctx.attr.target.label.name) != ctx.attr.name:
+        fail("target should follow `{java_library target name}-checkstyle` pattern")
+
+    properties = ctx.file.properties
+    suppressions = ctx.file.suppressions
+    opts = ctx.attr.opts
+    sopts = ctx.attr.string_opts
+
+    # Checkstyle and its dependencies
+    classpath = ":".join([file.path for file in ctx.files._classpath])
+
+    args = ""
+    inputs = []
+    if ctx.file.config:
+      args += " -c %s" % ctx.file.config.path
+      inputs.append(ctx.file.config)
+    if properties:
+      args += " -p %s" % properties.path
+      inputs.append(properties)
+    if suppressions:
+      inputs.append(suppressions)
+
+    # Build command to run Checkstyle test
+    cmd = " ".join(
+        ["java -cp %s com.puppycrawl.tools.checkstyle.Main" % classpath] +
+        [args] +
+        ["--%s" % x for x in opts] +
+        ["--%s %s" % (k, sopts[k]) for k in sopts] +
+        [file.path for file in ctx.attr.target[JavaSourceFiles].files]
+    )
+
+    # Wrap checkstyle command in a shell script so allow_failure is supported
+    ctx.actions.expand_template(
+        template = ctx.file._checkstyle_sh_template,
+        output = ctx.outputs.checkstyle_script,
+        substitutions = {
+            "{command}" : cmd,
+            "{allow_failure}": str(int(ctx.attr.allow_failure)),
+        },
+        is_executable = True,
+    )
+
+    files = [ctx.outputs.checkstyle_script, ctx.file.license] + ctx.attr.target[JavaSourceFiles].files + ctx.files._classpath + inputs
+    runfiles = ctx.runfiles(
+        files = files,
+        collect_data = True
+    )
+    return DefaultInfo(
+        executable = ctx.outputs.checkstyle_script,
+        files = depset(files),
+        runfiles = runfiles,
+    )
+
+checkstyle_test = rule(
+    implementation = _checkstyle_test_impl,
+    test = True,
+    attrs = {
+        "config": attr.label(
+            allow_single_file=True,
+            doc = "A checkstyle configuration file",
+            default = "//tools/checkstyle:config.xml",
+        ),
+        "suppressions": attr.label(
+            allow_single_file=True,
+            doc = "A checkstyle suppressions file"
+        ),
+        "license": attr.label(
+            allow_single_file=True,
+            doc = "A license file that can be used with the checkstyle license target",
+            default = "//tools/checkstyle:license-header.txt",
+        ),
+        "properties": attr.label(
+            allow_single_file=True,
+            doc = "A properties file to be used"
```

Comment:
What kind of file is this? What are the properties that it can contain? If you have an example, please provide one.

Reply:
@oferb I'm actually questioning removing the attribute completely as we don't use it, WDYT?
These properties, AFAIR, are the same you'd pass to `java` if you run it from cmdline (i.e. `-Xmx` for limiting memory allocation pool)

Reply:
Ok, so let's remove it.

Reply:
@oferb fixed in b1f8cbdeb78d5810a5d5827a1d198b3e65a7d851